profiling/src/collections/string_storage.rs

```diff
@@ -0,0 +1,208 @@
+use std::cell::Cell;
+use std::collections::HashMap;
+use std::hash::BuildHasherDefault;
+use std::ptr;
+use std::rc::Rc;
+
+use super::identifiable::FxIndexSet;
+use super::identifiable::Id;
+use super::identifiable::StringId;
+use super::string_table::StringTable;
+
+pub trait StringStorage {
+    /// Interns the `str` as a string, returning the id in the string table.
+    /// The empty string is guaranteed to have an id of [StringId::ZERO].
+    fn intern(&mut self, item: Rc<str>) -> StringId;
+    fn get_string(&self, id: StringId) -> Rc<str>;
+    fn len(&self) -> usize;
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+    fn into_iter(self: Box<Self>) -> Box<dyn Iterator<Item = Rc<str>>>;
+    fn clone_empty(&self) -> Box<dyn StringStorage>;
+}
+
+pub struct SimpleStringStorage {
+    set: FxIndexSet<Rc<str>>,
+}
+
+impl SimpleStringStorage {
+    pub fn new() -> Self {
+        SimpleStringStorage {
+            set: Default::default(),
+        }
+    }
+}
+
+impl Default for SimpleStringStorage {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl StringStorage for SimpleStringStorage {
+    fn intern(&mut self, item: Rc<str>) -> StringId {
+        // For performance, delay converting the [&str] to a [String] until
+        // after it has been determined to not exist in the set. This avoids
+        // temporary allocations.
+        let index = match self.set.get_index_of(&item) {
+            Some(index) => index,
+            None => {
+                let (index, _inserted) = self.set.insert_full(item.clone());
+                // This wouldn't make any sense; the item couldn't be found so
+                // we try to insert it, but suddenly it exists now?
+                debug_assert!(_inserted);
+                index
+            }
+        };
+        StringId::from_offset(index)
+    }
```

Comment:
It occurred to me that while ideally all this structure would be lock-free and whatnot, inserting and claiming an index is actually the part that would be really nice to be able to do in a non-blocking manner.

I was googling and didn't find a lot of interesting solid/production pre-existing data structure implementations with the kinds of characteristics we'd want (which is super disappointing -- I miss you java.util.concurrent), but yea, may be worth keeping an eye out if we find something.

Reply:
Maybe https://crates.io/crates/dashmap ? But we'd need to solve the index assignment (easy if we don't do index reuse, really annoying if we want to do that)